### PR TITLE
arch(rpc): narrow single-level positional gate to name[int] + migrate research parser behind row adapters (#1501 phase 1)

### PR DESCRIPTION
**Pull request by teng-lin**
## What & why

**Phase 1 of #1501** (drain the `SINGLE_LEVEL_ALLOWLIST` — the #1491 burndown ratchet). Two parts:

### A. Gate precision
The single-level detector flagged ANY `x[i]` int-literal subscript, mixing genuine RPC-payload reads with benign indexing. It now requires the subscripted value to be a bare `ast.Name` (`data[0]` / `result[2]` — the RPC-payload shape), excluding Attribute/Call-valued subscripts (`sys.version_info[0]`, `e.args[0]`, `.split(...)[0]`, `Path(...).parents[3]`).

**Verified safe across the whole feature tree**: all 15 non-`Name` flagged sites were benign — zero are RPC reads (a decoded batchexecute payload is always bound to a local before it is walked; both reviewers re-verified independently, including a `self.<attr>[int]` sweep that returned nothing). The new self-check documents this as a **convention-backed scope with named escapes** — a future `self._payload[0]` / `parse_rpc()[0]` idiom must widen the detector, not be adopted.

### B. First migration: the research parser
New `_row_adapters/research.py` (`ResearchTaskRow` — strict id/info via `safe_index`; `ResearchTaskInfoRow` + `ResearchResultRow` — soft length-guarded reads; `unwrap_poll_tasks` envelope probe), mirroring the chat/artifacts adapter conventions. `_research_task_parser.py` rewritten onto named properties with **exact soft-vs-strict semantics preserved** — a differential harness ran the verbatim originals against the new code over an edge-case battery: **0 mismatches**. 31 new adapter tests pin positions + behavior.

### Net
`SINGLE_LEVEL_ALLOWLIST`: **47 → 41** (5 benign-only files via gate precision + `_research_task_parser.py` via migration; 0 stale, 0 unbaselined — exact). Chained-descent gate untouched.

## Verification

- Full suite green (9401 passed); gate + stale-check + new self-checks pass; research parser/service/API tests unchanged (110 pass).
- mypy + ruff clean; API-compat audit **EXIT 0** (private-only changes); `docs/architecture.md` updated (freshness gate).
- Polished: claude (solid, hole-question empirically re-verified) + codex (**SHIP**; 3 Minor doc-accuracy fixes applied: POLL_RESEARCH id `e3bVqc`, "research parser" wording, convention-scoped detector docstring).

Part of #1501 (remaining ~41-file burndown continues there — do not auto-close).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Refactor**
  * Internal refactor of research/task parsing to use structured row adapters for more reliable, centralized parsing and fewer positional-index errors.

* **Documentation**
  * Architecture docs and file map updated to reflect the new adapter organization and structure.

* **Tests**
  * New unit tests validating research parsing behavior and envelope-unwrapping.
  * Tightened AST guardrail tests for single-level positional indexing detection.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_row_adapters/research.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        Descends through ``safe_index`` so an absent info slot RAISES — a task
+        row without its info block is genuine drift.
+        """
+        return safe_index(
+            self._raw, self._INFO_POS, method_id=_POLL_METHOD_ID, source=self._INFO_SOURCE
+        )
+
+
+@dataclass(frozen=True)
+class ResearchTaskInfoRow:
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_row_adapters/research.py]: Addressed in 57e621a8: removed the decorator — the class is ClassVar constants + staticmethods only, so frozen-dataclass machinery added nothing. Thanks!

File: src/notebooklm/_row_adapters/research.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        return query_info[ResearchTaskInfoRow._QUERY_TEXT_POS] if query_info else None
+
+    @staticmethod
+    def bundle_sources(bundle: Any) -> Any:
+        """Sources list at ``task_info[3][0]`` — the raw value (list-validated upstream).
+
+        ``bundle`` is ``task_info[3]`` (already validated as a non-empty list by
+        the caller), so ``[0]`` is always present.
+        """
+        return bundle[ResearchTaskInfoRow._SOURCES_POS]
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_row_adapters/research.py:181]: Addressed in d3582ed4: bundle_sources now length-guards to the missing-slot default (None, mirroring bundle_summary); the parser's caller coerces it to []. Thanks!

File: tests/unit/test_research_row_adapter.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        assert ResearchTaskInfoRow.query_text([]) is None
+
+    def test_bundle_sources_returns_first(self) -> None:
+        assert ResearchTaskInfoRow.bundle_sources([["a", "b"], "summary"]) == ["a", "b"]
+
+    def test_bundle_summary_present(self) -> None:
+        assert ResearchTaskInfoRow.bundle_summary([["a"], "Summary text"]) == "Summary text"
+
+    def test_bundle_summary_absent_returns_none(self) -> None:
+        assert ResearchTaskInfoRow.bundle_summary([["a"]]) is None
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/unit/test_research_row_adapter.py:121]: Addressed in d3582ed4: added test_bundle_sources_empty_bundle_soft_degrades pinning bundle_sources([]) is None.
